In [57]:
import pandas as pd

# Einlesen des Datensatzes

In [58]:
df = pd.read_csv('https://raw.githubusercontent.com/andreashrb/Python_for_ML_Project/main/data/copper_economic_enriched_week.csv', sep=',', encoding='utf-8', index_col=0, decimal=".", thousands=",", parse_dates=['date'])
df = df.reset_index()

In [59]:
# Anzeige der ersten 5 Zeilen des Datensatzes
df['id'] = 1
df.head()

date  lme_copper_cash  ...  Global_Verbrauch in 1000 Tonnen  id
0 2012-01-08          7574.75  ...                          14429.0   1
1 2012-01-15          7749.00  ...                          14429.0   1
2 2012-01-22          8188.60  ...                          14429.0   1
3 2012-01-29          8408.20  ...                          14429.0   1
4 2012-02-05          8379.00  ...                          14429.0   1

[5 rows x 32 columns]

In [60]:
# Anzeige Zeilen/Spalten des Datensatzes
df.shape

(470, 32)

In [61]:
# Ausgabe Datentypen der einzelnen Spalten des DataFrames
df.dtypes

date                                      datetime64[ns]
lme_copper_cash                                  float64
lme_copper_three_month                           float64
lme_copper_stock                                 float64
CN_BIP pro Kopf in USD                           float64
US_BIP pro Kopf in USD                           float64
Global_Nachfrage in Mio. Tonnen                  float64
EU_Produktionsmenge in 1000 Tonnen               float64
AU_Produktionsmenge in 1000 Tonnen               float64
CL_Produktionsmenge in 1000 Tonnen               float64
CN_Produktionsmenge in 1000 Tonnen               float64
MX_Produktionsmenge in 1000 Tonnen               float64
PE_Produktionsmenge in 1000 Tonnen               float64
RU_Produktionsmenge in 1000 Tonnen               float64
ZM_Produktionsmenge in 1000 Tonnen               float64
US_Produktionsmenge in 1000 Tonnen               float64
CN_Verbrauch in 1000 Tonnen                      float64
EU_Verbrauch in 1000 Tonnen    

# Feature Extraction mit tsfresh

In [62]:
# !pip install tsfresh
import tsfresh
from tsfresh import extract_features
from tsfresh import select_features
from tsfresh.utilities.dataframe_functions import impute
from tsfresh.utilities.dataframe_functions import roll_time_series

## Erstellung eines rollierenden Zeitreihendatensatzes als DataFrame

In [63]:
df_rolled = roll_time_series(df, column_id="id", column_sort="date", max_timeshift=10)

Rolling: 100%|█████████████████████████████████████████████████████████████████████████| 40/40 [00:06<00:00,  6.20it/s]


In [64]:
df_rolled.head(50)

date  lme_copper_cash  ...  Global_Verbrauch in 1000 Tonnen                        id
0  2012-01-08          7574.75  ...                          14429.0  (1, 2012-01-08 00:00:00)
1  2012-01-08          7574.75  ...                          14429.0  (1, 2012-01-15 00:00:00)
2  2012-01-15          7749.00  ...                          14429.0  (1, 2012-01-15 00:00:00)
3  2012-01-08          7574.75  ...                          14429.0  (1, 2012-01-22 00:00:00)
4  2012-01-15          7749.00  ...                          14429.0  (1, 2012-01-22 00:00:00)
5  2012-01-22          8188.60  ...                          14429.0  (1, 2012-01-22 00:00:00)
6  2012-01-08          7574.75  ...                          14429.0  (1, 2012-01-29 00:00:00)
7  2012-01-15          7749.00  ...                          14429.0  (1, 2012-01-29 00:00:00)
8  2012-01-22          8188.60  ...                          14429.0  (1, 2012-01-29 00:00:00)
9  2012-01-29          8408.20  ...                          14429.0  (1, 2012-01-29 00:00:00)
10 2012-01-08          7574.75  ...                          14429.0  (1, 2012-02-05 00:00:00)
11 2012-01-15          7749.00  ...                          14429.0  (1, 2012-02-05 00:00:00)
12 2012-01-22          8188.60  ...                          14429.0  (1, 2012-02-05 00:00:00)
13 2012-01-29          8408.20  ...                          14429.0  (1, 2012-02-05 00:00:00)
14 2012-02-05          8379.00  ...                          14429.0  (1, 2012-02-05 00:00:00)
15 2012-01-08          7574.75  ...                          14429.0  (1, 2012-02-12 00:00:00)
16 2012-01-15          7749.00  ...                          14429.0  (1, 2012-02-12 00:00:00)
17 2012-01-22          8188.60  ...                          14429.0  (1, 2012-02-12 00:00:00)
18 2012-01-29          8408.20  ...                          14429.0  (1, 2012-02-12 00:00:00)
19 2012-02-05          8379.00  ...                          14429.0  (1, 2012-02-12 00:00:00)
20 2012-02-12          8499.20  ...                          14429.0  (1, 2012-02-12 00:00:00)
21 2012-01-08          7574.75  ...                          14429.0  (1, 2012-02-19 00:00:00)
22 2012-01-15          7749.00  ...                          14429.0  (1, 2012-02-19 00:00:00)
23 2012-01-22          8188.60  ...                          14429.0  (1, 2012-02-19 00:00:00)
24 2012-01-29          8408.20  ...                          14429.0  (1, 2012-02-19 00:00:00)
25 2012-02-05          8379.00  ...                          14429.0  (1, 2012-02-19 00:00:00)
26 2012-02-12          8499.20  ...                          14429.0  (1, 2012-02-19 00:00:00)
27 2012-02-19          8378.40  ...                          14429.0  (1, 2012-02-19 00:00:00)
28 2012-01-08          7574.75  ...                          14429.0  (1, 2012-02-26 00:00:00)
29 2012-01-15          7749.00  ...                          14429.0  (1, 2012-02-26 00:00:00)
30 2012-01-22          8188.60  ...                          14429.0  (1, 2012-02-26 00:00:00)
31 2012-01-29          8408.20  ...                          14429.0  (1, 2012-02-26 00:00:00)
32 2012-02-05          8379.00  ...                          14429.0  (1, 2012-02-26 00:00:00)
33 2012-02-12          8499.20  ...                          14429.0  (1, 2012-02-26 00:00:00)
34 2012-02-19          8378.40  ...                          14429.0  (1, 2012-02-26 00:00:00)
35 2012-02-26          8359.80  ...                          14429.0  (1, 2012-02-26 00:00:00)
36 2012-01-08          7574.75  ...                          14429.0  (1, 2012-03-04 00:00:00)
37 2012-01-15          7749.00  ...                          14429.0  (1, 2012-03-04 00:00:00)
38 2012-01-22          8188.60  ...                          14429.0  (1, 2012-03-04 00:00:00)
39 2012-01-29          8408.20  ...                          14429.0  (1, 2012-03-04 00:00:00)
40 2012-02-05          8379.00  ...                          14429.0  (1, 2012-03-04 00:00:00)
41 2012-02-12      

In [65]:
df_rolled.shape

(5115, 32)

Der rollienrende Zeitreihendatensatz hat nun 110.685 Zeilen und weiterhin 32 Spalten. Aus 470 Zeilen wurden, durch den rollierenden Aufbau des Zeitreihendatensatzes, nun 110.685 Zeilen generiert. Sehr umfangreiche Zeitreihen werden auf diesem Weg sehr groß und könnten Probleme bei der Feature Extraction auf einer einfachen lokalen Maschine verursachen.

## Extraktion möglicher Features aus dem rollierenden Zeitreihendatensatz
Aus dem rollierend aufgebauten Zeitreihendatensatz (df_rolled) werden im folgenden alle möglichen Zeitreihenfeatures, die die Library tsfresh unterstützt, generiert.

In [66]:
# Extracting features from time series data
df_features = extract_features(df_rolled, column_id="id", column_sort="date", column_value='lme_copper_cash')

Feature Extraction: 100%|██████████████████████████████████████████████████████████████| 40/40 [00:10<00:00,  3.83it/s]


In [67]:
df_features2 = extract_features(df_rolled, column_id="id", column_sort="date", column_value='lme_copper_stock')

Feature Extraction: 100%|██████████████████████████████████████████████████████████████| 40/40 [00:10<00:00,  3.76it/s]


In [68]:
df_features.head()

lme_copper_cash__variance_larger_than_standard_deviation  ...  lme_copper_cash__mean_n_absolute_max__number_of_maxima_7
1 2012-01-08                                                0.0         ...                                                NaN       
  2012-01-15                                                1.0         ...                                                NaN       
  2012-01-22                                                1.0         ...                                                NaN       
  2012-01-29                                                1.0         ...                                                NaN       
  2012-02-05                                                1.0         ...                                                NaN       

[5 rows x 789 columns]

In [69]:
df_features2.head()

lme_copper_stock__variance_larger_than_standard_deviation  ...  lme_copper_stock__mean_n_absolute_max__number_of_maxima_7
1 2012-01-08                                                0.0          ...                                                NaN        
  2012-01-15                                                1.0          ...                                                NaN        
  2012-01-22                                                1.0          ...                                                NaN        
  2012-01-29                                                1.0          ...                                                NaN        
  2012-02-05                                                1.0          ...                                                NaN        

[5 rows x 789 columns]

In [70]:
df_features.shape

(470, 789)

In [71]:
df_features2.shape

(470, 789)

Es wurden 798 Features aus den vorliegenden Zeitreihen (Kupferpreis und Kupferreserven) extrahiert!

In [72]:
# Joinen der beiden Datensätze

df_features_final = pd.merge(df_features, df_features2, left_index=True, right_index=True)

In [73]:
df_features_final.head()

lme_copper_cash__variance_larger_than_standard_deviation  ...  lme_copper_stock__mean_n_absolute_max__number_of_maxima_7
1 2012-01-08                                                0.0         ...                                                NaN        
  2012-01-15                                                1.0         ...                                                NaN        
  2012-01-22                                                1.0         ...                                                NaN        
  2012-01-29                                                1.0         ...                                                NaN        
  2012-02-05                                                1.0         ...                                                NaN        

[5 rows x 1578 columns]

Insgesamt liegen nun 1578 Features zu den beiden Zeitreihen (Kupferpreis und Kupferreserven) vor.

## Selektion relevanter Features mit Bezug zur Zielvariable

In [74]:
# Harmonisierung der Indizierung des ursprünglichen DataFrames und Speicherung der Zielvariable "y"
df = df.set_index(['id', 'date'])
y = df['lme_copper_cash']
y

id  date      
1   2012-01-08    7574.75
    2012-01-15    7749.00
    2012-01-22    8188.60
    2012-01-29    8408.20
    2012-02-05    8379.00
                   ...   
    2020-12-06    7671.10
    2020-12-13    7689.80
    2020-12-20    7839.20
    2020-12-27    7787.25
    2021-01-03    7795.00
Name: lme_copper_cash, Length: 470, dtype: float64

In [75]:
# Aus den vorliegenden Features (789) werden jetzt die für die Vorhersage der Zielvariable potentiell relevanten Features herausgefiltert.

impute(df_features_final)
features_filtered = select_features(df_features_final, y)

C:\Users\woelk\anaconda3\envs\condatascience2\lib\site-packages\tsfresh\utilities\dataframe_functions.py:198: RuntimeWarning: The columns ['lme_copper_cash__partial_autocorrelation__lag_5'
 'lme_copper_cash__partial_autocorrelation__lag_6'
 'lme_copper_cash__partial_autocorrelation__lag_7'
 'lme_copper_cash__partial_autocorrelation__lag_8'
 'lme_copper_cash__partial_autocorrelation__lag_9'
 'lme_copper_cash__cwt_coefficients__coeff_11__w_2__widths_(2, 5, 10, 20)'
 'lme_copper_cash__cwt_coefficients__coeff_11__w_5__widths_(2, 5, 10, 20)'
 'lme_copper_cash__cwt_coefficients__coeff_11__w_10__widths_(2, 5, 10, 20)'
 'lme_copper_cash__cwt_coefficients__coeff_11__w_20__widths_(2, 5, 10, 20)'
 'lme_copper_cash__cwt_coefficients__coeff_12__w_2__widths_(2, 5, 10, 20)'
 'lme_copper_cash__cwt_coefficients__coeff_12__w_5__widths_(2, 5, 10, 20)'
 'lme_copper_cash__cwt_coefficients__coeff_12__w_10__widths_(2, 5, 10, 20)'
 'lme_copper_cash__cwt_coefficients__coeff_12__w_20__widths_(2, 5, 10, 20)'
 'l

In [76]:
features_filtered.head()

lme_copper_cash__cwt_coefficients__coeff_10__w_5__widths_(2, 5, 10, 20)  ...  lme_copper_stock__agg_linear_trend__attr_"rvalue"__chunk_len_10__f_agg_"mean"
1 2012-01-08                                        8602.117939                        ...                                               -1.0                            
  2012-01-15                                        8602.117939                        ...                                               -1.0                            
  2012-01-22                                        8602.117939                        ...                                               -1.0                            
  2012-01-29                                        8602.117939                        ...                                               -1.0                            
  2012-02-05                                        8602.117939                        ...                                               -1.0                            

[5 rows x 230 columns]

In [77]:
features_filtered.shape

(470, 230)

Aus den 1578 Features, die initial mit der tsfresh Library für die beiden Zeitreihen erzeugt und berechnet wurden, sind nach der Feature Selection noch 876 Features übrig geblieben.

Analog zum ursprünglichen DataFrame liegen wieder 470 Zeilen im Datensatz vor.

Lt. den Entwicklern der tsfresh-Library kann der DataFrame "features_filtered" nun verwendet werden, um ein Modell zu trainieren.

# Extraktion weiterer Features aus der Zeitreihe

Im Rahmen der Data Exploration wurde analysiert, inwiefern die Quartale, Monate, Kalenderwochen und Wochentage Einfluss auf den Kupferpreis haben. Letztlich ging es dabei primär um die Identifikation saisonaler Effekte, die sich auf den Kupferpreis und dessen Entwicklung auswirken. Gewisse saisonale Effekte sind, bei Betrachtung der gesamten Zeitreihe und der jeweiligen Durchschnittskupferpreise pro Quartal, Monat, Woche und Wochentag, nicht von der Hand zu weisen, auch wenn diese saisonalen Effekte, wie in der DataExploration bereits ausführlicher erläutert, zunächst einmal nicht besonders stark ausgefallen sind. Aus der Autokorrelationsanalyse heraus war sogar von fehlender statistischer Signifikanz die Rede. Dennoch sollen, aufgrund der leichten saisonalen Muster, die explorativ identifiziert werden konnten, auch diese Features wieder Berücksichtigung finden.

In [78]:
# Hinzufügen der Spalten "quarter", "month", "weekofyear" zum DataFrame "features_filtered"

df= df.reset_index()
df['quarter'] = pd.DatetimeIndex(df['date']).quarter
df['month'] = pd.DatetimeIndex(df['date']).month
df['weekofyear'] = pd.DatetimeIndex(df['date']).weekofyear
df['year'] = pd.DatetimeIndex(df['date']).year

# Setzen des Index, analog zum Dataframe "features_filtered"
df = df.set_index(['id','date'])

C:\Users\woelk\AppData\Local\Temp/ipykernel_12132/443322918.py:6: FutureWarning: weekofyear and week have been deprecated, please use DatetimeIndex.isocalendar().week instead, which returns a Series.  To exactly reproduce the behavior of week and weekofyear and return an Index, you may call pd.Int64Index(idx.isocalendar().week)
  df['weekofyear'] = pd.DatetimeIndex(df['date']).weekofyear


In [79]:
# Abspeichern der relevanten, zu joinenden, Features in separater Variable
df_to_join = df[['year','quarter','month','weekofyear']]

In [80]:
# Ausgabe des zu joinenden DataFrames
df_to_join.head()

year  quarter  month  weekofyear
id date                                        
1  2012-01-08  2012        1      1           1
   2012-01-15  2012        1      1           2
   2012-01-22  2012        1      1           3
   2012-01-29  2012        1      1           4
   2012-02-05  2012        1      2           5

In [81]:
# Für den Join auf den Indexspalten müssen diese identisch benannt sein
features_filtered = features_filtered.rename_axis(['id','date'])

In [82]:
# Join
features_filtered = pd.merge(features_filtered, df_to_join, left_index=True, right_index=True)

In [83]:
# Ausgabe des gejointen DataFrames
features_filtered.head()

lme_copper_cash__cwt_coefficients__coeff_10__w_5__widths_(2, 5, 10, 20)  ...  weekofyear
id date                                                                                 ...            
1  2012-01-08                                        8602.117939                        ...           1
   2012-01-15                                        8602.117939                        ...           2
   2012-01-22                                        8602.117939                        ...           3
   2012-01-29                                        8602.117939                        ...           4
   2012-02-05                                        8602.117939                        ...           5

[5 rows x 234 columns]

In [84]:
from tsfresh.feature_selection.relevance import calculate_relevance_table

relevance_table = calculate_relevance_table(features_filtered, y)
relevance_table = relevance_table[relevance_table.relevant]
relevance_table.sort_values('p_value', inplace = True)

relevance_table.head(877)

feature  ... relevant
feature                                                                                                ...         
lme_copper_cash__cwt_coefficients__coeff_10__w_...  lme_copper_cash__cwt_coefficients__coeff_10__w...  ...     True
lme_copper_cash__cwt_coefficients__coeff_9__w_5...  lme_copper_cash__cwt_coefficients__coeff_9__w_...  ...     True
lme_copper_cash__cwt_coefficients__coeff_10__w_...  lme_copper_cash__cwt_coefficients__coeff_10__w...  ...     True
lme_copper_cash__cwt_coefficients__coeff_9__w_2...  lme_copper_cash__cwt_coefficients__coeff_9__w_...  ...     True
lme_copper_cash__cwt_coefficients__coeff_10__w_...  lme_copper_cash__cwt_coefficients__coeff_10__w...  ...     True
...                                                                                               ...  ...      ...
lme_copper_stock__autocorrelation__lag_7                     lme_copper_stock__autocorrelation__lag_7  ...     True
lme_copper_cash__change_quantiles__f_agg_"var"_...  lme_copper_cash__change_quantiles__f_agg_"var"...  ...     True
lme_copper_stock__cwt_coefficients__coeff_4__w_...  lme_copper_stock__cwt_coefficients__coeff_4__w...  ...     True
lme_copper_stock__approximate_entropy__m_2__r_0.7   lme_copper_stock__approximate_entropy__m_2__r_0.7  ...     True
lme_copper_stock__agg_linear_trend__attr_"rvalu...  lme_copper_stock__agg_linear_trend__attr_"rval...  ...     True

[231 rows x 4 columns]

In [90]:
df_selected = pd.DataFrame()

for i in relevance_table['feature']:
    df_selected[i] = features_filtered[i]
    print(i)

lme_copper_cash__cwt_coefficients__coeff_10__w_5__widths_(2, 5, 10, 20)
lme_copper_cash__cwt_coefficients__coeff_9__w_5__widths_(2, 5, 10, 20)
lme_copper_cash__cwt_coefficients__coeff_10__w_10__widths_(2, 5, 10, 20)
lme_copper_cash__cwt_coefficients__coeff_9__w_2__widths_(2, 5, 10, 20)
lme_copper_cash__cwt_coefficients__coeff_10__w_20__widths_(2, 5, 10, 20)
lme_copper_cash__cwt_coefficients__coeff_8__w_5__widths_(2, 5, 10, 20)
lme_copper_cash__cwt_coefficients__coeff_9__w_10__widths_(2, 5, 10, 20)
lme_copper_cash__cwt_coefficients__coeff_9__w_20__widths_(2, 5, 10, 20)
lme_copper_cash__cwt_coefficients__coeff_8__w_10__widths_(2, 5, 10, 20)
lme_copper_cash__cwt_coefficients__coeff_7__w_5__widths_(2, 5, 10, 20)
lme_copper_cash__minimum
lme_copper_cash__mean
lme_copper_cash__root_mean_square
lme_copper_cash__cwt_coefficients__coeff_8__w_20__widths_(2, 5, 10, 20)
lme_copper_cash__absolute_maximum
lme_copper_cash__maximum
lme_copper_cash__cwt_coefficients__coeff_8__w_2__widths_(2, 5, 10, 20)

C:\Users\woelk\AppData\Local\Temp/ipykernel_12132/1037737054.py:4: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`
  df_selected[i] = features_filtered[i]


In [91]:
df_selected.head()

lme_copper_cash__cwt_coefficients__coeff_10__w_5__widths_(2, 5, 10, 20)  ...  lme_copper_stock__agg_linear_trend__attr_"rvalue"__chunk_len_10__f_agg_"mean"
id date                                                                                 ...                                                                               
1  2012-01-08                                        8602.117939                        ...                                               -1.0                            
   2012-01-15                                        8602.117939                        ...                                               -1.0                            
   2012-01-22                                        8602.117939                        ...                                               -1.0                            
   2012-01-29                                        8602.117939                        ...                                               -1.0                            
   2012-02-05                                        8602.117939                        ...                                               -1.0                            

[5 rows x 231 columns]

In [92]:
df_selected.shape

(470, 231)

# Speichern der Datensätze

In [93]:
# Speichern der Input-Daten (features_filtered)
df_selected.to_csv('C:\\Users\\woelk\\OneDrive\\Dokumente\\Studium - HdM Stuttgart\\Machine Learning\\0 - Projekt\\Kupferpreisprognose\\prepared_input_data.csv', sep=",", encoding = 'utf-8')

In [94]:
# Umwandeln der Output-Daten in einen DataFrame & Speichern der Output-Daten (y)

df_y= pd.DataFrame(y)
df_y.to_csv('C:\\Users\\woelk\\OneDrive\\Dokumente\\Studium - HdM Stuttgart\\Machine Learning\\0 - Projekt\\Kupferpreisprognose\\output_data.csv', sep=",", encoding = 'utf-8')

In [98]:
# Joinen von x und y
final_df = pd.merge(df_selected, df_y,left_index=True, right_index=True)
final_df.to_csv('C:\\Users\\woelk\\OneDrive\\Dokumente\\Studium - HdM Stuttgart\\Machine Learning\\0 - Projekt\\Kupferpreisprognose\\input_output_data.csv', sep=",", encoding = 'utf-8')
